### PR TITLE
generic: Perform partial evaluation during constant propagation

### PR DESCRIPTION
**Pull request by IagoAbal**
So that pattern `foo(1)` will match:

    y = 1
    x = y + 0
    foo(x)

and pattern `foo(True)` will match:

    y = True
    x = y or False
    foo(x)

and pattern `foo("hello world")` will match:

    y = "hello"
    x = y + " world"
    foo(x)

A small step towards returntocorp/semgrep#1645!

### CODE REVIEW

**Review by aryx — APPROVED** [review]
Looks great.
Could be nice after to add unit tests in semgrep-core/tests/.../
with the examples in your commit message.

**Comment by aryx** [comment]
@DrewDennison iago PR are still not validated by circleCI. Normal?

File: lang_GENERIC/analyze/Constant_propagation.ml [hunk 1]
```diff
@@ -120,6 +128,101 @@ let var_stats prog : var_stats =
   visitor (Pr prog);
   h
 
+(*****************************************************************************)
+(* Partial evaluation *)
+(*****************************************************************************)
+
+let literal_of_bool b =
+  let b_str = string_of_bool b in
+  let tok   = Parse_info.fake_info b_str in
```
[the remark this answers is not in the document]
Comment (IagoAbal) [reply lang_GENERIC/analyze/Constant_propagation.ml:137]: I will fix that! Is it preferable to use the token of e.g. `+` or is it better to still use a `FakeTokStr` but fill in the "next to" field to refer to the `+`?

File: lang_GENERIC/analyze/Constant_propagation.ml [hunk 2]
```diff
@@ -120,6 +128,101 @@ let var_stats prog : var_stats =
...
+        Some (literal_of_int i)
+    | Minus, Some (Int _ as l) ->
+        int_of_literal l >>= fun i ->
+        Some (literal_of_int (-i))
+    | __else__ -> None
+    )
+  | [Arg e1; Arg e2] ->
+    (match eval_expr env e1, eval_expr env e2 with
+    | Some (Bool _ as l1), Some (Bool _ as l2) ->
+        bool_of_literal l1 >>= fun b1 ->
```
[the remark this answers is not in the document]
Comment (IagoAbal) [reply lang_GENERIC/analyze/Constant_propagation.ml:208]: Thanks for pointing it out! I was somewhat aware, but I haven't yet used many of the recent features of the language (I'm kind of a slow adopter). I'll give the `let*` a try.

File: lang_GENERIC/analyze/Constant_propagation.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aryx) [reply lang_GENERIC/analyze/Constant_propagation.ml:137]: use an existing token (e.g, +), I almost never use the next-to field of FakeTokStr (it's used mostly with C/C++ to partially expand some macros to be able to parse code)

File: lang_GENERIC/analyze/Constant_propagation.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (aryx) [reply lang_GENERIC/analyze/Constant_propagation.ml:208]: I'm a slow adopter too :)  I just recently look at all the recent features added to OCaml in the last few years (GADT, attributes, extension points, monadic let, etc.). There's a few good stuff in it.

**Pull request merged**